src/Servers/Kestrel/Transport.Quic/src/Internal/QuicConnectionListener.cs

```diff
@@ -134,8 +140,14 @@ public async ValueTask CreateListenerAsync()
         {
             var quicConnection = await _listener.AcceptConnectionAsync(cancellationToken);
 
-            // _currentAcceptingConnection is set inside ConnectionOptionsCallback.
-            var connectionContext = _currentAcceptingConnection;
+            if (!_pendingConnections.TryGetValue(quicConnection, out var connectionContext))
+            {
+                throw new InvalidOperationException("Couldn't find ConnectionContext for QuicConnection.");
```

Comment:
What's keeping the `QuicConnectionContext` rooted until we get to this point? It has no more strong references as soon as `await _tlsConnectionCallbackOptions.OnConnection(context, cancellationToken);` completes, right?

Reply:
The CWT is keeping `QuicConnectionContext` rooted. The important thing to keep rooted so the CWT isn't emptied is `QuicConnection`.

A couple of lines above this:

```
var quicConnection = await _listener.AcceptConnectionAsync(cancellationToken);
```

Reply:
Right. I forgot how CWT works lol.